backend/stakepoold/rpc/rpcserver/client.go

```diff
@@ -0,0 +1,168 @@
+// that client through a mutex.
+//
+// It should be noted that this is a temporary fix to the problem that rpcclient
+// does not return an error when autoreconnect is turned on but the client is
+// disconnected. The permanent solution is to change the behaviour of rpccleint.
+// TODO: Remove this file.
+type Client struct {
+	// client is protected by a mutex that must be held for reads/writes.
+	client *rpcclient.Client
+	mux    sync.RWMutex
+
+	cfg          *rpcclient.ConnConfig
+	ntfnHandlers *rpcclient.NotificationHandlers
+	wg           sync.WaitGroup
+	stop         chan struct{}
+
+	connected    chan struct{}
+	connectedMux sync.Mutex
+}
+
+// Connected returns a receiving copy of the current connected channel. If
+// disconnected and the channel is not yet blocking, creates a new channel that
+// will be closed on a successful reconnect.
+func (c *Client) Connected() <-chan struct{} {
+	c.connectedMux.Lock()
+	defer c.connectedMux.Unlock()
+	if c.RPCClient().Disconnected() {
+		// Start blocking on connected chan if not already.
+		select {
+		case <-c.connected:
+			c.connected = make(chan struct{})
+		default:
+		}
+	}
+	return c.connected
+}
+
+// RPCClient allows access to the underlying rpcclient by providing a copy of
+// its address.
+func (c *Client) RPCClient() *rpcclient.Client {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return c.client
+}
+
+// New creates a new Client and starts the automatic reconnection handler.
+// Returns an error if unable to construct a new rpcclient.
+func New(cfg *rpcclient.ConnConfig, ntfnHandlers *rpcclient.NotificationHandlers) (*Client, error) {
+	client, err := rpcclient.New(cfg, ntfnHandlers)
+	if err != nil {
+		return nil, err
+	}
+	c := &Client{
+		client:       client,
+		cfg:          cfg,
+		ntfnHandlers: ntfnHandlers,
+		stop:         make(chan struct{}),
+		connected:    make(chan struct{}),
+	}
+	// A closed connected channel indcates successfully connected.
+	close(c.connected)
+	go func(c *Client) {
+		c.wg.Add(1)
+		c.autoReconnect()
+	}(c)
+	return c, nil
+}
+
+// Stop stops automatic reconnections.
+func (c *Client) Stop() {
+	c.stop <- struct{}{}
+	c.wg.Wait()
+	// Stop blocking on connected if blocking, as we will never reconnect again.
+	select {
+	case <-c.connected:
+	default:
+		close(c.connected)
+	}
+}
+
+// autoReconnect waits for a disconnect or stop. On disconnect it attempts to
+// reconnect to the client every connectionRetryInterval.
+//
+// This function must be run as a goroutine.
+func (c *Client) autoReconnect() {
+out:
+	for {
+		select {
+		case <-c.stop:
+			break out
+		case <-time.After(disconnectCheckInterval):
+			if !c.RPCClient().Disconnected() {
+				continue out
+			}
+			// Client is disconnected. Try to reconnect.
+			// Start blocking on connected chan if not already.
+			c.connectedMux.Lock()
+			select {
+			case <-c.connected:
+				c.connected = make(chan struct{})
+			default:
+			}
+			c.connectedMux.Unlock()
```

Comment:
For the above dozen or so lines, much of `Connected` is duplicated, so why not do instead:

```go
select {
case <-c.Connected():   // not connected
default:                // still connected
	continue
}
```

Although, the above demonstrates that the semantics of `Connected()` are somewhat confusing, so maybe consider a `func IsConnected() bool`.